src/grpc_server.cc

```diff
@@ -4299,8 +4299,10 @@ ModelStreamInferHandler::StreamInferResponseComplete(
                  << state->cb_count_ << ", flags " << flags;
 
 #ifdef TRITON_ENABLE_TRACING
-  state->trace_timestamps_.emplace_back(std::make_pair(
-      "INFER_RESPONSE_COMPLETE", TraceManager::CaptureTimestamp()));
+  if (state->cb_count_ == 1) {
```

Comment:
I guess we don't really formulate how tracing will work on decoupled model? And if we are only recording it once, does it make sense to record on the last response complete?

Reply:
The intent of INFER_RESPONSE_COMPLETE flag was to record when the callback was invoked to test our preserve_ordering logic by skipping the uncertainties added by our networking libraries.

Coming back to your question, you are correct that we have not formulated when to record INFER_RESPONSE_COMPLETE. There are three options:

1.  **Record at the first callback**: No need of mutex locks to protect trace_timestamps_. The other timestamps will be recorded after the first callback one by one.
2.  **Record every callback**: trace_timestamps_ needs protection that might affect performance.
3.  **Record at the final callback**: trace_timestamps_ will still need protection, as GRPC_SEND_START or GRPC_SEND_END of previous would still be in the process of being recorded when this timestamp is added.

In short, by recording the timestamp just at the first response callback we are able to prevent mutex locks and I think user would not be more concerned about INFER_RESPONSE_COMPLETE anyways. They might be more interested in GRPC_SEND_START, which in most likelihood should be very close to INFER_RESPONSE_COMPLETE.

I can go with any other option if you would like.

Reply:
I was just curious why recording at the first callback doesn't require mutex locks. Is there a possibility of the second infer response arriving before the first infer response (maybe in the non-preserve ordering case)?

Reply:
`state->cb_count_` is incremented in L4294. And for decoupled we require that only one response is sent at a time. 
Hence, the first `StreamInferResponseComplete` call will increment the cb_count_ and no other responses for this state object are not into GRPC_SEND* cycle.